File: backend/src/main/java/project/backend/domain/member/app/MemberService.java
package project.backend.domain.member.app;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import project.backend.domain.imagefile.ImageFileService;
import project.backend.domain.member.dao.MemberRepository;
import project.backend.domain.member.dto.PasswordChangeRequest;
import project.backend.domain.member.dto.event.ProfileUpdateEvent;
import project.backend.auth.dto.MemberDetails;
import project.backend.domain.member.dto.MemberResponse;
import project.backend.domain.member.dto.MemberInfoUpdateRequest;
import project.backend.domain.member.dto.SignUpRequest;
import project.backend.domain.member.entity.Member;
import project.backend.domain.member.entity.ProviderType;
import project.backend.domain.member.mapper.MemberMapper;
import project.backend.global.exception.errorcode.AuthErrorCode;
import project.backend.global.exception.errorcode.MemberErrorCode;
import project.backend.global.exception.ex.AuthException;
import project.backend.global.exception.ex.MemberException;

@Slf4j
@Service
@Transactional
@RequiredArgsConstructor
public class MemberService {

	private final MemberRepository memberRepository;
	private final ImageFileService imageFileService;
	private final PasswordEncoder passwordEncoder;
	private final ApplicationEventPublisher eventPublisher;

	@Value("${file.images.profile.default}")
	private String defaultProfileImg;

	public MemberResponse saveMember(SignUpRequest request) {

		if (checkUsernameAlreadyExists(request.getUsername())) {
			throw new MemberException(MemberErrorCode.USERNAME_ALREADY_EXISTS);
		}

		if (request.getEmail() != null && checkEmailAlreadyExists(request.getEmail())) {
			throw new MemberException(MemberErrorCode.EMAIL_ALREADY_EXISTS);
		}

		String encryptedPassword = passwordEncoder.encode(request.getPassword());

		Member newMember = memberRepository.save(
			MemberMapper.toEntity(request, encryptedPassword, defaultProfileImg));

		return MemberMapper.toResponse(newMember);
	}

	public MemberResponse updateMemberInfo(Authentication auth, MemberInfoUpdateRequest request) {
		MemberDetails memberDetails = (MemberDetails) auth.getPrincipal();
		Member targetMember = getMemberById(memberDetails.getId());

		doUpdateMemberInfo(targetMember, request);

		eventPublisher.publishEvent(ProfileUpdateEvent.of(targetMember));

		return MemberMapper.toResponse(targetMember);
	}

	private void doUpdateMemberInfo(Member targetMember, MemberInfoUpdateRequest request) {
		if (request.nickname() != null) {
			targetMember.updateNickname(request.nickname());
		}

		if (request.email() != null) {
			targetMember.updateEmail(request.email());
		}

		if (request.profileImg() != null) {
			String profileImage = imageFileService.saveProfileImage(request.profileImg());
			targetMember.updateProfileImage(profileImage);
		}
	}

	public void updatePassword(Authentication auth, PasswordChangeRequest request) {
		MemberDetails memberDetails = (MemberDetails) auth.getPrincipal();
		Member targetMember = getMemberById(memberDetails.getId());

		if (targetMember.getProvider() != ProviderType.LOCAL) {
			throw new AuthException(AuthErrorCode.WRONG_AUTH_TYPE_LOGIN);
		}

		String currentPassword = targetMember.getPassword();

		if (!passwordEncoder.matches(request.currentPassword(),
			currentPassword)) {
			throw new MemberException(MemberErrorCode.WRONG_PASSWORD);
		}

		if (passwordEncoder.matches(request.newPassword(), currentPassword)) {
			throw new MemberException(MemberErrorCode.SAME_AS_OLD_PASSWORD);
		}

		targetMember.updatePassword(request.newPassword(), passwordEncoder);
	}


	// Spring Security에서 UsernameNotFoundException을 처리하도록 유도하는 메서드
	public Member getMemberForLogin(String username) {
		try {
			return getMemberByUsername(username);
		} catch (MemberException e) {
			log.info("존재하지 않는 username으로 로그인 시도: {}", username);
			throw new UsernameNotFoundException("존재하지 않는 유저입니다: " + username, e);
		}
	}

	public Member getMemberByUsername(String username) {
		return memberRepository.findByUsername(username)
			.orElseThrow(() -> new MemberException(MemberErrorCode.MEMBER_NOT_FOUND));
	}

	public Member getMemberById(Long id) {
		return memberRepository.findById(id)
			.orElseThrow(() -> new MemberException(MemberErrorCode.MEMBER_NOT_FOUND));
	}

	public MemberResponse getMemberDetails(Authentication auth) {
		MemberDetails loginMember = (MemberDetails) auth.getPrincipal();
		Long memberId = loginMember.getId();
		log.info("memberId = {}", memberId);
		Member member = getMemberById(memberId);
		return MemberMapper.toResponse(member);
	}

	private boolean checkUsernameAlreadyExists(String username) {
		return memberRepository.findByUsername(username).isPresent();
	}

	private boolean checkEmailAlreadyExists(String email) {
		return memberRepository.findByEmail(email).isPresent();
	}

}

File: backend/src/main/java/project/backend/domain/member/app/MemberService.kt
package project.backend.domain.member.app

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.ApplicationEventPublisher
import org.springframework.security.core.Authentication
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile
import project.backend.auth.dto.MemberDetails
import project.backend.domain.imagefile.ImageFileService
import project.backend.domain.member.dao.MemberRepository
import project.backend.domain.member.dto.MemberInfoUpdateRequest
import project.backend.domain.member.dto.MemberResponse
import project.backend.domain.member.dto.PasswordChangeRequest
import project.backend.domain.member.dto.SignUpRequest
import project.backend.domain.member.dto.event.ProfileUpdateEvent.Companion.of
import project.backend.domain.member.entity.Member
import project.backend.domain.member.entity.ProviderType
import project.backend.domain.member.mapper.MemberMapper.toEntity
import project.backend.domain.member.mapper.MemberMapper.toResponse
import project.backend.global.exception.errorcode.AuthErrorCode
import project.backend.global.exception.errorcode.MemberErrorCode
import project.backend.global.exception.ex.AuthException
import project.backend.global.exception.ex.MemberException

@Service
@Transactional
class MemberService (
    private val memberRepository: MemberRepository,
    private val imageFileService: ImageFileService,
    private val passwordEncoder: PasswordEncoder,
    private val eventPublisher: ApplicationEventPublisher,

){
    @Value("\${file.images.profile.default}")
    private lateinit var defaultProfileImg: String

    fun saveMember(request: SignUpRequest): MemberResponse {
        if (checkUsernameAlreadyExists(request.username)) {
            throw MemberException(MemberErrorCode.USERNAME_ALREADY_EXISTS)
        }

        if (request.email != null && checkEmailAlreadyExists(request.email)) {
            throw MemberException(MemberErrorCode.EMAIL_ALREADY_EXISTS)
        }

        val encryptedPassword = passwordEncoder.encode(request.password)

        val newMember = memberRepository.save(
            toEntity(request, encryptedPassword, defaultProfileImg)
        )

        return toResponse(newMember)
    }

    fun updateMemberInfo(
        auth: Authentication, request: MemberInfoUpdateRequest,
        file: MultipartFile?
    ): MemberResponse {
        val memberDetails = auth.principal as MemberDetails
        val targetMember = getMemberById(memberDetails.id)

        doUpdateMemberInfo(targetMember, request, file)

        eventPublisher.publishEvent(of(targetMember))

        return toResponse(targetMember)
    }

    private fun doUpdateMemberInfo(
        targetMember: Member, request: MemberInfoUpdateRequest,
        file: MultipartFile?
    ) {
        if (request.nickname != null) {
            targetMember.updateNickname(request.nickname)
        }

        if (request.email != null) {
            targetMember.updateEmail(request.email)
        }

        if (file != null) {
            val profileImage = imageFileService.saveProfileImage(file)
            targetMember.updateProfileImage(profileImage)
        }
    }

    fun updatePassword(auth: Authentication, request: PasswordChangeRequest) {
        val memberDetails = auth.principal as MemberDetails
        val targetMember = getMemberById(memberDetails.id)

        if (targetMember.provider != ProviderType.LOCAL) {
            throw AuthException(AuthErrorCode.WRONG_AUTH_TYPE_LOGIN)
        }

        val currentPassword = targetMember.password

        if (!passwordEncoder.matches(
                request.currentPassword,
                currentPassword
            )
        ) {
            throw MemberException(MemberErrorCode.WRONG_PASSWORD)
        }

        if (passwordEncoder.matches(request.newPassword, currentPassword)) {
            throw MemberException(MemberErrorCode.SAME_AS_OLD_PASSWORD)
        }

        targetMember.updatePassword(request.newPassword, passwordEncoder)
    }


    // Spring Security에서 UsernameNotFoundException을 처리하도록 유도하는 메서드
    fun getMemberForLogin(username: String): Member {
        try {
            return getMemberByUsername(username)
        } catch (e: MemberException) {
//            MemberService.log.info("존재하지 않는 username으로 로그인 시도: {}", username)
            throw UsernameNotFoundException("존재하지 않는 유저입니다: $username", e)
        }
    }

    fun getMemberByUsername(username: String): Member {
        return memberRepository.findByUsername(username)
            ?: throw MemberException(MemberErrorCode.MEMBER_NOT_FOUND)
    }

    fun getMemberById(id: Long): Member =
        memberRepository.findById(id)
            .orElseThrow { MemberException(MemberErrorCode.MEMBER_NOT_FOUND) }

    fun getMemberDetails(auth: Authentication): MemberResponse {
        val loginMember = auth.principal as MemberDetails
        val memberId = loginMember.id
//        MemberService.log.info("memberId = {}", memberId)
        val member = getMemberById(memberId)
        return toResponse(member)
    }

    private fun checkUsernameAlreadyExists(username: String): Boolean =
        memberRepository.findByUsername(username) != null

    private fun checkEmailAlreadyExists(email: String): Boolean =
        memberRepository.findByEmail(email) != null
}
